pkg/dyninst/symdb/symdb.go

```diff
@@ -0,0 +1,775 @@
+	// the top of this stack.
+	blockStack []codeBlock
+}
+
+// PCRange corresponds to a range of program counters (PCs). Ranges are
+// inclusive of the start and exclusive of the end.
+type PCRange = [2]uint64
+
+// codeBlock abstracts LexicalBlocks, Subprograms and InlinedSubroutines. They
+// all correspond to a list of PC ranges.
+type codeBlock interface {
+	resolvePCRanges(dwarfData *dwarf.Data) ([]PCRange, error)
+}
+
+// dwarfBlock is the implementation of codeBlock for lexical blocks.
+type dwarfBlock struct {
+	// The DWARF entry for the lexical block.
+	entry    *dwarf.Entry
+	pcRanges []PCRange
+	// pcRangesResolved is set if pcRanges has been already calculated.
+	pcRangesResolved bool
+}
+
+// dwarfBlock implements codeBlock.
+var _ codeBlock = (*dwarfBlock)(nil)
+
+// resolvePCRanges parses and memoizes the ranges attribute of the dwarf entry.
+func (d *dwarfBlock) resolvePCRanges(dwarfData *dwarf.Data) ([]PCRange, error) {
+	if d.pcRangesResolved {
+		return d.pcRanges, nil
+	}
+
+	ranges, err := dwarfData.Ranges(d.entry)
+	if err != nil {
+		return nil, err
+	}
+	d.pcRanges = ranges
+	d.pcRangesResolved = true
+	return d.pcRanges, nil
+}
+
+// resolveLine computes the start and end lines of the lexical block by
+// combining PC ranges from DWARF with pclinetab data. Returns [0,0] if the
+// block has no address ranges, and thus no lines.
+func (d *dwarfBlock) resolveLines(dwarfData *dwarf.Data, funcTable *gosym.Func) ([2]int, error) {
+	// Find the start and end lines of the block by going through the address
+	// ranges, resolving them to lines, and selecting the minimum and maximum.
+	ranges, err := d.resolvePCRanges(dwarfData)
+	if err != nil {
+		return [2]int{}, err
+	}
+	if len(ranges) == 0 {
+		return [2]int{}, nil
+	}
+	startLine := math.MaxInt
+	endLine := 0
+	for _, r := range ranges {
+		rangeStartLine := funcTable.LineTable.PCToLine(r[0])
+		rangeEndLine := funcTable.LineTable.PCToLine(r[1])
+		if rangeStartLine < startLine {
+			startLine = rangeStartLine
+		}
+		if rangeEndLine > endLine {
+			endLine = rangeEndLine
+		}
+	}
+	return [2]int{startLine, endLine}, nil
+}
+
+// subprogBlock is the implementation of codeBlock for subprograms. The block's
+// PC ranges are resolved trivially to a single range.
+type subprogBlock struct {
+	lowpc  uint64
+	highpc uint64
+}
+
+// subprogBlock implements codeBlock.
+var _ codeBlock = subprogBlock{}
+
+// resolvePCRanges implements the codeBlock interface.
+func (s subprogBlock) resolvePCRanges(*dwarf.Data) ([]PCRange, error) {
+	return []PCRange{{s.lowpc, s.highpc}}, nil
+}
+
+func NewSymDBBuilder(
+	dwarfData *dwarf.Data, sym *gosym.Table, loclistReader *object.LoclistReader, pointerSize int,
+) *SymDBBuilder {
+	return &SymDBBuilder{
+		dwarfData:     dwarfData,
+		sym:           sym,
+		loclistReader: loclistReader,
+		// TODO: Should we use entryReader.AddressSize() as pointerSize?
+		pointerSize: pointerSize,
```

Comment:
What's used now? I think the pointer size (8) should be used.

Reply:
I forwarded the definition of Architecture into `object` and it has a `PointerSize()` method. You'll be using the `object.File` anyway I hope, so just use that.

Reply:
switched to using `object.ElfFile`. Why do we need the `object.File` interface?